prov/efa/src/efa_cq.c

```diff
@@ -248,16 +250,38 @@ static struct fi_ops efa_cq_fi_ops = {
 	.ops_open = fi_no_ops_open,
 };
 
-
 /**
- * @brief Create and set cq->ibv_cq_ex
+ * @brief Create ibv_cq_ex by calling ibv_create_cq_ex
  *
- * @param[in] cq Pointer to the efa_cq. cq->ibv_cq_ex must be NULL.
+ * @param[in] cq Pointer to the efa_cq.
  * @param[in] attr Pointer to fi_cq_attr.
- * @param[out] Return code = 0 if successful, or negative otherwise.
+ * @param[out] Return pointer to ibv_cq_ex if successful, otherwise NULL
  */
+static inline struct ibv_cq_ex *efa_cq_ibv_create_cq_ex(struct efa_cq *cq, struct fi_cq_attr *attr) {
+	struct ibv_cq_init_attr_ex init_attr_ex = {
+		.cqe = attr->size ? attr->size : EFA_DEF_CQ_SIZE,
+		.cq_context = NULL,
+		.channel = NULL,
+		.comp_vector = 0,
+		/* EFA requires these values for wc_flags and comp_mask.
+		 * See `efa_create_cq_ex` in rdma-core.
+		 */
+		.wc_flags = IBV_WC_STANDARD_FLAGS,
+		.comp_mask = 0,
+	};
+
+	return ibv_create_cq_ex(cq->domain->device->ibv_ctx, &init_attr_ex);
+}
+
 #if HAVE_EFADV_CQ_EX
-static inline int efa_cq_set_ibv_cq_ex(struct efa_cq *cq, struct fi_cq_attr *attr) {
+/**
+ * @brief Create ibv_cq_ex by calling efadv_create_cq
+ *
+ * @param[in] cq Pointer to the efa_cq.
+ * @param[in] attr Pointer to fi_cq_attr.
+ * @param[out] Return pointer to ibv_cq_ex if successful, otherwise NULL
+ */
+static inline struct ibv_cq_ex *efa_cq_efadv_create_cq(struct efa_cq *cq, struct fi_cq_attr *attr) {
```

Comment:
I am not sure we need this function.

Cann't we just call `efadv_create_cq` in the HAVE_EFADV_CREATE_CQ section of `efa_cq_set_ibv_cq_ex`?

Reply:
Sounds good. Will address.

Reply:
Addressed